File: backend/src/main/java/org/globe42/dao/UserDaoImpl.java
package org.globe42.dao;

import java.util.List;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

/**
 * Implementation of {@link UserDaoCustom}
 * @author JB Nizet
 */
public class UserDaoImpl implements UserDaoCustom {

    @PersistenceContext
    private EntityManager em;

    @Override
    public boolean existsNotDeletedById(Long userId) {
        return existsByQueryAndId("select 1 from User u where u.id = :id and u.deleted = false",
                                  userId);
    }

    @Override
    public boolean existsNotDeletedAdminById(Long userId) {
        return existsByQueryAndId("select 1 from User u where u.id = :id and u.deleted = false and u.admin = true",
                                  userId);
    }

    private boolean existsByQueryAndId(String jpql, Long userId) {
        List<Number> result = em.createQuery(jpql, Number.class)
                                .setParameter("id", userId)
                                .getResultList();
        return !result.isEmpty();
    }
}

File: backend/src/main/kotlin/org/globe42/dao/UserDaoImpl.kt
package org.globe42.dao

import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

/**
 * Implementation of [UserDaoCustom]
 * @author JB Nizet
 */
class UserDaoImpl : UserDaoCustom {

    @PersistenceContext
    private lateinit var em: EntityManager

    override fun existsNotDeletedById(userId: Long) =
            existsByQueryAndId("select 1 from User u where u.id = :id and u.deleted = false", userId)

    override fun existsNotDeletedAdminById(userId: Long) =
            existsByQueryAndId("select 1 from User u where u.id = :id and u.deleted = false and u.admin = true",
                               userId)

    private fun existsByQueryAndId(jpql: String, userId: Long?): Boolean {
        val result = em.createQuery(jpql, Number::class.java)
                .setParameter("id", userId)
                .resultList
        return !result.isEmpty()
    }
}
